causal_testing/surrogate/surrogate_search_algorithms.py

```diff
@@ -0,0 +1,224 @@
+        rel_split = str(relationship).split(" ")
+
+        if rel_split[1] == ">=":
+            var_space[rel_split[0]]["low"] = int(rel_split[2])
+        elif rel_split[1] == "<=":
+            var_space[rel_split[0]]["high"] = int(rel_split[2])
+
+    gene_space = []
+    gene_space.append(var_space[surrogate.treatment])
+    for adj in surrogate.adjustment_set:
+        gene_space.append(var_space[adj])
+
+    gene_types = []
+    gene_types.append(scenario.variables.get(surrogate.treatment).datatype)
+    for adj in surrogate.adjustment_set:
+        gene_types.append(scenario.variables.get(adj).datatype)
+
+    ga = GA(
+        num_generations=200,
+        num_parents_mating=4,
+        fitness_func=build_fitness_func(surrogate, delta),
+        sol_per_pop=10,
+        num_genes=1 + len(surrogate.adjustment_set),
+        gene_space=gene_space,
+        gene_type=gene_types,
+    )
+
+    if config is not None:
+        for k, v in config.items():
+            if k == "gene_space":
+                raise Exception(
+                    "Gene space should not be set through config. This is generated from the causal specification"
+                )
+            setattr(ga, k, v)
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        ga.run()
+    solution, fitness, _idx = ga.best_solution()
+
+    solution_dict = dict()
+    solution_dict[surrogate.treatment] = solution[0]
+    for idx, adj in enumerate(surrogate.adjustment_set):
+        solution_dict[adj] = solution[idx + 1]
+
+    return (solution_dict, fitness, surrogate)
+
+
+class MultiProcessGeneticSearchAlgorithm(SearchAlgorithm):
+    def __init__(self, delta=0.05, config: dict = None, processes: int = 1) -> None:
+        self.delta = delta
+        self.config = config
+        self.processes = processes
+
+    def generate_fitness_functions(self, surrogate_models: list[Estimator]) -> list[SearchFitnessFunction]:
+        return [SearchFitnessFunction(None, model) for model in surrogate_models]
+
+    def search(self, fitness_functions: list[SearchFitnessFunction], specification: CausalSpecification) -> list:
+        global pool_vals
+        solutions = []
+        all_fitness_function = fitness_functions.copy()
+
+        while len(all_fitness_function) > 0:
+            num = self.processes
+            if num > len(all_fitness_function):
+                num = len(all_fitness_function)
+
+            pool_vals.clear()
+
+            while len(pool_vals) < num and len(all_fitness_function) > 0:
+                fitness_function = all_fitness_function.pop()
+                pool_vals.append(
+                    (fitness_function.surrogate_model, self.delta, specification.scenario, self.config)
+                )
+
+            with mp.Pool(processes=num) as pool:
+                solutions.extend(pool.map(threaded_search_function, range(len(pool_vals))))
+
+        return min(
+            solutions, key=itemgetter(1)
```

Comment:
Another `itemgetter` I'm unsure of if it is working as intended

Reply:
Same as above